### PR TITLE
Add flag controlling creation of `.packages` file.

### PR DESCRIPTION
**Pull request by lrhn**
WDYT?

Related to https://github.com/dart-lang/pub/issues/2756

### CODE REVIEW

**Comment by jonasfj** [comment]
Hmm, I think we have to update some of the code that asserts state is up-to-date.

You're suggesting, we: add a flag that allows disabling generation of `.packages` and we later flip the default for the flag.

I can see the argument for starting to remove `.packages` now that we're doing null-safety.

**Comment by sigurdm** [comment]
Hmm, I think we have to update some of the code that asserts state is up-to-date.

I have landed: https://github.com/dart-lang/pub/pull/2764 that removes the assert that .packages is up to date in case it doesn't exist, it also removes several other places where we wrongly relied on .packages.

**Comment by mit-mit** [comment]
@lrhn @sigurdm do we still need this?

**Comment by sigurdm** [comment]
Hmm - I don't remember where we ended up last we discussed this. 

But IMO we should just continue creating this file until dart 3.
No reason to make unnecessary breaking changes before a breaking version shift.
On the other hand we could generate a comment inside the .packages file explaining that it is deprecated.

**Comment by sigurdm** [comment]
I merged this with upstream and turned the flag to be off by default.

**Comment by sigurdm** [comment]
ping @jonasfj

**Comment by mit-mit** [comment]
@sigurdm can you clarify the flag behavoir? Is the flag `--packages-file` & off by default?

**Comment by sigurdm** [comment]
> Is the flag --packages-file to create the .packages file 

yes

> it's off by default (i.e. when this lands we'll no longer create the file by default)?

yes

**Comment by mit-mit** [comment]
OK. We would have to send the breaking change notice for this first then.

This is a slightly faster approach than the one discussed in https://github.com/dart-lang/pub/issues/2756 (which suggested first having a flag you must opt-in to using to skip the generation of .packages), but given how old this file is, I'm OK with moving ahead here. @lrhn WDYT?

**Comment by sigurdm** [comment]
My thinking was that now we had the deprecation notice in the file we should be able to take this step.

But I'm also ok with having the .packages generation still be on by default.

**Comment by lrhn** [comment]
We introduced `package_config.json` in Dart 2.8. I think almost two years should be enough warning!

File: lib/src/command/add.dart [hunk 1]
```diff
@@ -92,6 +92,8 @@ class AddCommand extends PubCommand {
         help: 'Build executables in immediate dependencies.');
     argParser.addOption('directory',
         abbr: 'C', help: 'Run this in the directory <dir>.', valueHelp: 'dir');
+    argParser.addFlag('packages-file',
+        help: 'Generate the legacy ".packages" file');
```
Comment (jonasfj) [inline lib/src/command/add.dart:96]: Should we specify `negatable: false` ? So we don't have `--no-packages-file` option?

File: test/testdata/goldens/help_test/pub upgrade --help.txt [hunk 2]
```diff
@@ -5,16 +5,18 @@ $ pub upgrade --help
...
--C, --directory=<dir>    Run this in the directory<dir>.
+-h, --help                  Print this usage information.
+    --[no-]offline          Use cached packages instead of accessing the
+                            network.
+-n, --dry-run               Report what dependencies would change but don't
+                            change any.
+    --[no-]precompile       Precompile executables in immediate dependencies.
+    --null-safety           Upgrade constraints in pubspec.yaml to null-safety
+                            versions
+    --[no-]packages-file    Generate the legacy ".packages" file
```
Comment (jonasfj) [inline test/testdata/goldens/help_test/pub upgrade --help.txt:16]: Let's just have `--packages-file`, by specifying `negatable: false`

**Review by jonasfj — APPROVED** [review]

File: lib/src/command/add.dart (hunk 1 above)
Comment (sigurdm) [reply lib/src/command/add.dart:96]: Yeah - let's do that

File: test/testdata/goldens/help_test/pub upgrade --help.txt (hunk 2 above)
Comment (sigurdm) [reply test/testdata/goldens/help_test/pub upgrade --help.txt:16]: Done

File: lib/src/command/add.dart [hunk 3]
```diff
@@ -93,7 +93,7 @@ class AddCommand extends PubCommand {
     argParser.addOption('directory',
         abbr: 'C', help: 'Run this in the directory <dir>.', valueHelp: 'dir');
     argParser.addFlag('packages-file',
```
Comment (mit-mit) [inline lib/src/command/add.dart:95]: Would it make sense to cal this flag `legacy-package-file` just to be extraordinarily clear?

**Pull request merged**